src/main/kotlin/org/rust/lang/core/types/infer/MemoryCategorization.kt

```diff
@@ -5,57 +5,167 @@
 
 package org.rust.lang.core.types.infer
 
+import com.intellij.psi.util.parentOfType
 import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsElement
 import org.rust.lang.core.psi.ext.containerExpr
 import org.rust.lang.core.psi.ext.mutability
 import org.rust.lang.core.types.builtinDeref
-import org.rust.lang.core.types.infer.Adjustment.Deref
+import org.rust.lang.core.types.infer.BorrowKind.ImmutableBorrow
+import org.rust.lang.core.types.infer.BorrowKind.MutableBorrow
+import org.rust.lang.core.types.infer.Categorization.*
+import org.rust.lang.core.types.infer.ImmutabilityBlame.*
+import org.rust.lang.core.types.infer.InteriorKind.InteriorElement
+import org.rust.lang.core.types.infer.InteriorKind.InteriorField
+import org.rust.lang.core.types.infer.MutabilityCategory.Declared
+import org.rust.lang.core.types.infer.PointerKind.BorrowedPointer
+import org.rust.lang.core.types.infer.PointerKind.UnsafePointer
 import org.rust.lang.core.types.inference
 import org.rust.lang.core.types.isDereference
+import org.rust.lang.core.types.regions.ReStatic
+import org.rust.lang.core.types.regions.Region
 import org.rust.lang.core.types.ty.*
 import org.rust.lang.core.types.type
 import org.rust.stdext.nextOrNull
 
 
+/** [Categorization] is a subset of the full expression forms */
+sealed class Categorization {
+    /** Temporary value */
+    data class Rvalue(val region: Region) : Categorization()
+
+    /** Static value */
+    object StaticItem : Categorization()
+
+    /** Variable captured by closure */
+    object Upvar : Categorization()
```

Comment:
Object?

Reply:
In rustc, `Upvar` contains `ClosureKind` (`Fn`, `FnMut`, and `FnOnce`, which are not currently supported in our type inference as far as I understand) and `UpvarId` (it would be the pair of the original `RsElement` and closure, but I'm not sure it would be required). So for now I keep this class empty; do you prefer `class Upvar() : Categorization()` instead of `object`?

Reply:
Yes, if we need to distinguish between `Upvar` instances (e.g., to store them in a `Set`).

Reply:
OK, it has a point.